File: testdata/fixtures/go/CWE-20.go
// CWE-20: Improper input validation — query param reaches exec/query without validation.
package main

import (
    "database/sql"
    "net/http"
    "os/exec"
)

func vulnExec(w http.ResponseWriter, r *http.Request) {
    q := r.URL.Query().Get("q")
    // ruleid: go-input-validation
    exec.Command("sh", "-c", q)
}

func vulnQuery(w http.ResponseWriter, r *http.Request, db *sql.DB) {
    id := r.URL.Query().Get("id")
    // ruleid: go-input-validation
    db.Query("SELECT * FROM users WHERE id = " + id)
}

func vulnExec2(w http.ResponseWriter, r *http.Request) {
    cmd := r.URL.Query().Get("cmd")
    // ruleid: go-input-validation
    exec.Command(cmd)
}

func safeFixed(w http.ResponseWriter, r *http.Request) {
    // ok: go-input-validation
    exec.Command("ls")
    // ok: go-input-validation
    db, _ := sql.Open("sqlite3", ":memory:")
    db.Query("SELECT * FROM users WHERE id = 1")
}

func safeValidated(w http.ResponseWriter, r *http.Request) {
    q := r.URL.Query().Get("q")
    if q != "allowed" {
        http.Error(w, "bad", 400)
        return
    }
    // ok: go-input-validation
    exec.Command("echo", "fixed")
}

File: testdata/fixtures/python/CWE-20.py
# CWE-20: Improper input validation — input() without validation reaches sink.
import sqlite3

def vuln_eval():
    x = input()
    # ruleid: python-input-validation
    eval(x)

def vuln_exec():
    data = input("enter: ")
    # ruleid: python-input-validation
    exec(data)

def vuln_sql():
    user = input()
    conn = sqlite3.connect("db.sqlite")
    cur = conn.cursor()
    # ruleid: python-input-validation
    cur.execute("SELECT * FROM users WHERE name = '" + user + "'")

def safe_regex():
    x = input()
    import re
    if not re.match(r"^[a-z]+$", x):
        raise ValueError("bad")
    # ok: python-input-validation
    eval("fixed_string")

def safe_no_input():
    y = "fixed"
    # ok: python-input-validation
    eval(y)
    # ok: python-input-validation
    cur = sqlite3.connect("a").cursor()
    cur.execute("SELECT * FROM t WHERE id=1")

def vuln_request_sql():
    from flask import request
    name = request.args.get("name")
    cur = sqlite3.connect("db.sqlite").cursor()
    # ruleid: python-input-validation
    cur.execute("SELECT * FROM users WHERE name = '" + name + "'")

def vuln_request_open():
    from flask import request
    path = request.form["path"]
    # ruleid: python-input-validation
    open(path)

def safe_request_literal():
    from flask import request
    _ = request.args.get("ignored")
    # ok: python-input-validation
    open("/etc/hostname")
